### PR TITLE
Add `FullNodeClient` to tests where `GatewayClient` is tested

### LINKED ISSUE #920: Test thoroughly FullNodeClient
In many tests, GatewayClient is used as *the* client. It should be changed to test both clients.

### PR DESCRIPTION
**Pull request by tomek0123456789**
Closes #920

## Introduced changes

- Changed implementation of tests to test not only GatewayClient, but also FullNodeClient

Mostly changed fixtures passed to tests from `gateway_client` and `gateway_account` to `client` and `account` respectively (these two return also the FullNode one). Sometimes tests needed more specific parametrization, like [here](https://github.com/software-mansion/starknet.py/blob/feat/full-node-clients-tests/starknet_py/tests/e2e/client/client_test.py#L208).

##

Issues that arose from this PR:
- #984 
- #985

### CODE REVIEW

File: starknet_py/tests/e2e/docs/guide/test_handling_client_errors.py [hunk 1]
```diff
@@ -11,7 +11,6 @@ async def test_handling_client_errors(gateway_account):
     try:
         contract_address = "1"  # Doesn't exist
         # docs: end
-        account = gateway_account
         # docs: start
```
Comment (drknzz) [inline starknet_py/tests/e2e/docs/guide/test_handling_client_errors.py:14]: The `# docs` comments are obsolete now, that nothing is between them

File: starknet_py/tests/e2e/deploy/deployer_test.py [hunk 2]
```diff
@@ -94,6 +94,7 @@ async def test_constructor_arguments_contract_deploy(
     )
 
 
+# TODO when changed to `account`, tests (1, True) and (2, False) fail on full_node_account
```
Comment (drknzz) [inline starknet_py/tests/e2e/deploy/deployer_test.py:97]: Could you perform some further investigation?
Is this a bug is FullNodeClient or?
What's the error message?

File: starknet_py/tests/e2e/account/account_test.py [hunk 3]
```diff
@@ -25,10 +25,10 @@
 
 @pytest.mark.run_on_devnet
 @pytest.mark.asyncio
-async def test_get_balance_throws_when_token_not_specified(account):
+async def test_get_balance_throws_when_token_not_specified(account, client):
     modified_account = Account(
         address=account.address,
-        client=GatewayClient(net="custom.net"),
+        client=client,
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:31]: I think we can try doing something like 

`@pytest.mark.parametrize("client_class", (GatewayClient, FullNodeClient))`

File: starknet_py/tests/e2e/account/account_test.py [hunk 4]
```diff
@@ -233,6 +232,8 @@ async def test_sign_declare_v2_transaction(
     assert signed_tx.max_fee == MAX_FEE
 
 
+# TODO full_node_account doesn't work here because declare_v2 isn't supported, change was introduced in RPC v0.3.0
+# and devnet hasn't been updated yet
 @pytest.mark.asyncio
 async def test_sign_declare_v2_transaction_auto_estimate(
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:238]: I think it's unlikely we'll have 0.3.0 support in starknet.py before this is merged.

[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/tests/e2e/account/account_test.py:238]: I'd crate an issue for that before merge.

File: starknet_py/tests/e2e/account/account_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomek0123456789) [reply starknet_py/tests/e2e/account/account_test.py:31]: oops, i knew i forgot something about one test, thanks😅

File: starknet_py/tests/e2e/account/account_test.py [hunk 5]
```diff
@@ -381,20 +389,36 @@ async def test_deploy_account_raises_on_no_enough_funds(deploy_account_details_f
...
+    [
+        (
+            "starknet_py.net.gateway_client.GatewayClient.call_contract",
+            "starknet_py.net.gateway_client.GatewayClient.deploy_account",
+            "gateway_client",
+        ),
+        (
+            "starknet_py.net.full_node_client.FullNodeClient.call_contract",
+            "starknet_py.net.full_node_client.FullNodeClient.deploy_account",
+            "full_node_client",
```
Comment (cptartur) [inline starknet_py/tests/e2e/account/account_test.py:410]: We don't need both `call_contract` and `deploy_account` here. Just base paths to gateway and full node client, specific methods can be then constructed with `+ ".call_contract"` etc.

File: starknet_py/tests/e2e/client/client_test.py [hunk 6]
```diff
@@ -204,31 +204,58 @@ async def test_get_class_by_hash(client, class_hash):
...
+            "full_node_client",
+            "starknet_py.net.full_node_client.FullNodeClient.get_transaction_receipt",
+        ),
+        (
+            "gateway_client",
+            "starknet_py.net.gateway_client.GatewayClient.get_transaction_receipt",
+        ),
+    ],
+)
+async def test_wait_for_tx_accepted(client, get_tx_receipt, request):
```
Comment (cptartur) [inline starknet_py/tests/e2e/client/client_test.py:220]: One possible problem I'm unsure of. If `client` is already a parametrized fixture, won't this cause this test to run for all client x path combinations e.g. 
GatewayClient and "starknet_py.net.full_node_client.FullNodeClient.get_transaction_receipt"?

cc @drknzz

File: starknet_py/tests/e2e/client/client_test.py [hunk 7]
```diff
@@ -204,31 +204,58 @@ async def test_get_class_by_hash(client, class_hash):
...
-        "starknet_py.net.gateway_client.GatewayClient.get_transaction_receipt",
+        get_tx_receipt,
         AsyncMock(),
     ) as mocked_receipt:
         mocked_receipt.return_value = TransactionReceipt(
             hash=0x1, status=TransactionStatus.ACCEPTED_ON_L2, block_number=1
         )
-
-        block_number, tx_status = await gateway_client.wait_for_tx(tx_hash=0x1)
+        client = request.getfixturevalue(client)
```
Comment (cptartur) [inline starknet_py/tests/e2e/client/client_test.py:228]: Why was this added?

File: starknet_py/tests/e2e/client/client_test.py [hunk 8]
```diff
@@ -248,34 +275,62 @@ async def test_wait_for_tx_pending(gateway_client):
...
     with patch(
-        "starknet_py.net.gateway_client.GatewayClient.get_transaction_receipt",
+        get_tx_receipt,
         AsyncMock(),
     ) as mocked_receipt:
         mocked_receipt.return_value = TransactionReceipt(
             hash=0x1, status=TransactionStatus.PENDING, block_number=1
         )
-
+        client = request.getfixturevalue(client)
```
Comment (cptartur) [inline starknet_py/tests/e2e/client/client_test.py:331]: Same here

File: starknet_py/tests/e2e/docs/code_examples/test_contract.py [hunk 9]
```diff
@@ -23,7 +23,7 @@ def test_init():
         ],
         provider=Account(
             address=0x321,
-            client=GatewayClient(TESTNET),
```
Comment (cptartur) [inline starknet_py/tests/e2e/docs/code_examples/test_contract.py:26]: I'm not sure about defaulting to full node client in docs already.

cc @war-in @drknzz

File: starknet_py/net/account/account_test.py [hunk 10]
```diff
@@ -12,8 +13,18 @@
...
+    [
+        "starknet_py.net.gateway_client.GatewayClient.call_contract",
+        "starknet_py.net.full_node_client.FullNodeClient.call_contract",
+    ],
+)
+async def test_get_balance_default_token_address(net, call_contract):
+    if "gateway" in call_contract:
+        client = GatewayClient(net=net)
+    else:
+        client = FullNodeClient(node_url=net + "/rpc")
```
Comment (cptartur) [inline starknet_py/net/account/account_test.py:27]: I don't really like `if else` like this in tests but alternative would be duplication I think.

Wdyt @drknzz @war-in

File: starknet_py/tests/e2e/client/client_test.py (hunk 6 above)
Comment (drknzz) [reply starknet_py/tests/e2e/client/client_test.py:220]: I was not sure about that too. I've checked the CI, and only 2 tests with that name have been ran 👍

File: starknet_py/tests/e2e/docs/code_examples/test_contract.py (hunk 9 above)
Comment (drknzz) [reply starknet_py/tests/e2e/docs/code_examples/test_contract.py:26]: Right, it should be a separate PR, after 0.16, to default to FullNodeClient in the docs.

File: starknet_py/net/account/account_test.py (hunk 10 above)
Comment (drknzz) [reply starknet_py/net/account/account_test.py:27]: I think it's the best we can get

Comment (war-in) [reply starknet_py/net/account/account_test.py:27]: How about that?

```
client = (
        GatewayClient(net=net)
        if "gateway" in call_contract
        else FullNodeClient(net=net + "/rpc")
    )
```

I know it is still `if/else` but in a nicer format

File: starknet_py/tests/e2e/client/client_test.py [hunk 11]
```diff
@@ -338,6 +393,49 @@ async def test_custom_session(map_contract, network):
...
+    assert full_node_client1._client.session.closed is False
+    assert full_node_client2._client.session.closed is False
+
+    await session.close()
+
+    assert full_node_client1._client.session.closed is True
+    assert full_node_client2._client.session.closed is True
+
+
+# TODO
```
[the remark this answers is not in the document]
Comment (tomek0123456789) [reply starknet_py/tests/e2e/client/client_test.py:438]: Uhhh, didn't put a space after the TODO and pycharm didn't detect it, my bad

File: starknet_py/tests/e2e/account/account_test.py [hunk 12]
```diff
@@ -381,20 +389,36 @@ async def test_deploy_account_raises_on_no_enough_funds(deploy_account_details_f
...
+        ),
+    ],
+)
+async def test_deploy_account_passes_on_enough_funds(
+    deploy_account_details_factory, call_contract, deploy_account, client, request
+):
     address, key_pair, salt, class_hash = await deploy_account_details_factory.get()
+    client = request.getfixturevalue(client)
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_test.py:418]: Why do we need this here? Wouldn't it be enough to pass client fixture?

File: starknet_py/tests/e2e/client/client_test.py (hunk 7 above)
Comment (tomek0123456789) [reply starknet_py/tests/e2e/client/client_test.py:228]: Because we don't want to mix GatewayClient methods with FullNodeClient instance (the topic you mentioned above this thread).

File: starknet_py/tests/e2e/account/account_test.py (hunk 12 above)
Comment (tomek0123456789) [reply starknet_py/tests/e2e/account/account_test.py:418]: Without it `client` fixture would be used as:
- FullNodeClient - full_node method
- FullNodeClient - gateway method
- GatewayClient - full_node method
- GatewayClient - gateway method.

File: starknet_py/tests/e2e/client/client_test.py [hunk 13]
```diff
@@ -338,6 +393,49 @@ async def test_custom_session(map_contract, network):
...
+    gateway2_response = await full_node_client2.get_transaction_receipt(tx_hash=tx_hash)
+    assert gateway1_response == gateway2_response
+
+    assert full_node_client1._client.session.closed is False
+    assert full_node_client2._client.session.closed is False
+
+    await session.close()
+
+    assert full_node_client1._client.session.closed is True
+    assert full_node_client2._client.session.closed is True
```
[the remark this answers is not in the document]
Comment (tomek0123456789) [reply starknet_py/tests/e2e/client/client_test.py:435]: I'm not sure about that, FullNodeClient has `_client property`, Gateway has `_feeder_gateway_client`, so asserts still need to stay. Refactoring in order to extract the other parts isn't worth it IMO since we want to deprecate Gateway later on anyway.

[the remark this answers is not in the document]
Comment (drknzz) [reply starknet_py/tests/e2e/client/client_test.py:435]: It's 40 lines of almost duplicate code, if we can get rid of it, we would want to :)

How about

```
internal_client = client._client if isinstance(client, FullNodeClient) else client._feeder_gateway_client
session = internal_client.session
```

or

```
internal_client = getattr(client, "_client", client._feeder_gateway_client)
```

though I'm unsure about the second one working, might need 2x getattr.

File: starknet_py/tests/e2e/deploy/deployer_test.py (hunk 2 above)
Comment (tomek0123456789) [reply starknet_py/tests/e2e/deploy/deployer_test.py:97]: In both cases, the transaction gets rejected with an unknown error, so I don't think it is an issue with FullNodeClient
```
salt = 1, pass_account_address = True
account = <starknet_py.net.account.account.Account object at 0x116a7e250>
map_class_hash = 269454390811152316...6789178223652524401

    @pytest.mark.asyncio
    @pytest.mark.parametrize(
        "salt, pass_account_address", [(1, True), (2, False), (None, True), (None, False)]
    )
    async def test_address_computation(
        salt, pass_account_address, account, map_class_hash
    ):
        deployer = Deployer(
            account_address=account.address if pass_account_address else None
        )
    
        deploy_call, computed_address = deployer.create_deployment_call(
            class_hash=map_class_hash, salt=salt
        )
    
        deploy_invoke_tx = await account.sign_invoke_transaction(
            deploy_call, max_fee=MAX_FEE
        )
        resp = await account.client.send_transaction(deploy_invoke_tx)
>       await account.client.wait_for_tx(resp.transaction_hash)

deployer_test.py:117: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <starknet_py.net.full_node_client.FullNodeClient object at 0x116a7edc0>
tx_hash = 159922644870949472...4208212143205555769, wait_for_accept = False
check_interval = 5

    async def wait_for_tx(
        self,
        tx_hash: Hash,
        wait_for_accept: Optional[bool] = False,
        check_interval=5,
    ) -> Tuple[int, TransactionStatus]:
        # pylint: disable=too-many-branches
        """
        Awaits for transaction to get accepted or at least pending by polling its status
    
        :param tx_hash: Transaction's hash
        :param wait_for_accept: If true waits for at least ACCEPTED_ON_L2 status, otherwise waits for at least PENDING
        :param check_interval: Defines interval between checks
        :return: Tuple containing block number and transaction status
        """
        if check_interval <= 0:
            raise ValueError("Argument check_interval has to be greater than 0.")
    
        first_run = True
        try:
            while True:
                result = await self.get_transaction_receipt(tx_hash=tx_hash)
                status = result.status
    
                if status in (
                    TransactionStatus.ACCEPTED_ON_L1,
                    TransactionStatus.ACCEPTED_ON_L2,
                ):
                    assert result.block_number is not None
                    return result.block_number, status
                if status == TransactionStatus.PENDING:
                    if not wait_for_accept:
                        if result.block_number is not None:
                            return result.block_number, status
                elif status == TransactionStatus.REJECTED:
>                   raise TransactionRejectedError(
                        message=result.rejection_reason,
E                       starknet_py.transaction_exceptions.TransactionRejectedError: Transaction was rejected with following Starknet error: Unknown Starknet error..

../../../net/client.py:169: TransactionRejectedError
```

Comment (drknzz) [reply starknet_py/tests/e2e/deploy/deployer_test.py:97]: Something similar was happening in #975 which has been fixed by altering the constructor arguments.

I'd suggest try changing something between the set of tests for GatewayClient and FullNodeClient (map_contract does not have a constructor, so maybe try chaning the max_fee? if that does not work maybe try with a contract that has a constructor and do something similar to #975).

Not exactly sure what causes the problem atm.
We probably would want to investigate this further, might be a devnet issue.

Comment (tomek0123456789) [reply starknet_py/tests/e2e/deploy/deployer_test.py]: Salt is now changed when`full_node_account` is tested, transactions with identical parameters won't be accepted I guess.

File: starknet_py/tests/e2e/docs/code_examples/test_contract.py (hunk 9 above)
Comment (tomek0123456789) [reply starknet_py/tests/e2e/docs/code_examples/test_contract.py]: Created an issue targeting the problem.

File: starknet_py/tests/e2e/account/account_test.py [hunk 14]
```diff
@@ -25,10 +26,16 @@
...
 @pytest.mark.run_on_devnet
 @pytest.mark.asyncio
-async def test_get_balance_throws_when_token_not_specified(account):
+@pytest.mark.parametrize("client_class", [GatewayClient, FullNodeClient])
+async def test_get_balance_throws_when_token_not_specified(account, client_class):
+    client = (
+        GatewayClient(net="custom.net")
+        if client_class is GatewayClient
+        else FullNodeClient(node_url="custom.net/rpc")
+    )
```
Comment (cptartur) [inline starknet_py/tests/e2e/account/account_test.py:35]: You're not really using the class reference from `client_class` beyond creating the client, so I'd either change it to 
```py
client = (
        client_class(net="custom.net")
        if client_class is GatewayClient
        else client_class(node_url="custom.net/rpc")
    )
```

or modify the parametrize entirely

```py
@pytest.mark.parametrize("client_class", ["GatewayClient, "FullNodeClient"])
async def test_get_balance_throws_when_token_not_specified(account, client_class):
    client = (
        GatewayClient(net="custom.net")
        if client_class == "GatewayClient"
        else FullNodeClient(node_url="custom.net/rpc")
    )
```

Not sure which approach is better, but I'd choose one for consistency.

File: starknet_py/tests/e2e/account/account_test.py [hunk 15]
```diff
@@ -233,6 +239,8 @@ async def test_sign_declare_v2_transaction(
     assert signed_tx.max_fee == MAX_FEE
 
 
+# TODO full_node_account doesn't work here because declare_v2 isn't supported, change was introduced in RPC v0.3.0
+# and devnet hasn't been updated yet
```
Comment (cptartur) [inline starknet_py/tests/e2e/account/account_test.py:243]: If we have an issue open for this, we could reference its number in the TODO.

File: starknet_py/tests/e2e/client/client_test.py [hunk 16]
```diff
@@ -315,27 +371,45 @@ async def test_custom_session(map_contract, network):
...
+    client1 = (
+        GatewayClient(net=network, session=session)
+        if isinstance(client_class, GatewayClient)
+        else FullNodeClient(node_url=network + "/rpc", net=network, session=session)
+    )
+    client2 = (
+        GatewayClient(net=network, session=session)
+        if isinstance(client_class, GatewayClient)
+        else FullNodeClient(node_url=network + "/rpc", net=network, session=session)
+    )
```
Comment (cptartur) [inline starknet_py/tests/e2e/client/client_test.py:383]: Same as the comment above

**Review by cptartur — COMMENTED** [review]
I left some comments, but it looks good overall.

**Review by drknzz — APPROVED** [review]
Great job 👍 

Please edit the description of the PR to include a short summary of what has changed (rather of what is left to change; also link any issues that arised from this PR)

File: starknet_py/tests/e2e/deploy/deployer_test.py [hunk 17]
```diff
@@ -98,26 +98,26 @@ async def test_constructor_arguments_contract_deploy(
...
-    salt, pass_account_address, gateway_account, map_class_hash
-):
+async def test_address_computation(salt, pass_account_address, account, map_class_hash):
     deployer = Deployer(
-        account_address=gateway_account.address if pass_account_address else None
+        account_address=account.address if pass_account_address else None
     )
+    if isinstance(salt, int) and isinstance(account.client, FullNodeClient):
+        # transactions have to be different for each account
+        salt = salt + 1
```
Comment (war-in) [inline starknet_py/tests/e2e/deploy/deployer_test.py:107]:
```suggestion
        salt += 2
```
One of the `salt` is `1`, so after adding 1 it will be `2` as a second case

File: starknet_py/tests/e2e/account/account_test.py [hunk 18]
```diff
@@ -26,13 +26,11 @@
...
-    client = (
-        client_class(net="custom.net")
-        if client_class is GatewayClient
-        else client_class(node_url="custom.net/rpc")
-    )
+@pytest.mark.parametrize(
+    "client",
+    [GatewayClient(net="custom.net"), FullNodeClient(node_url="custom.net/rpc")],
+)
+async def test_get_balance_throws_when_token_not_specified(account, client):
```
Comment (war-in) [inline starknet_py/tests/e2e/account/account_test.py:33]: ❤️

File: starknet_py/tests/e2e/fixtures/misc.py [hunk 19]
```diff
@@ -73,6 +73,16 @@ def typed_data(request) -> TypedData:
     return typed_data
 
 
+@pytest.fixture(name="tx_receipt_full_node_path")
+def get_tx_receipt_full_node_client():
+    return "starknet_py.net.full_node_client.FullNodeClient.get_transaction_receipt"
+
+
+@pytest.fixture(name="tx_receipt_gateway_path")
```
Comment (war-in) [inline starknet_py/tests/e2e/fixtures/misc.py:81]: please add scope 🙏

Comment (war-in) [reply starknet_py/tests/e2e/fixtures/misc.py:81]: To both of them

**Review by war-in — APPROVED** [review]
